iothub_client/inc/iothub_device_client.h

```diff
@@ -349,6 +348,88 @@ extern "C"
     */
     MOCKABLE_FUNCTION(, IOTHUB_CLIENT_RESULT, IoTHubDeviceClient_UploadMultipleBlocksToBlobAsync, IOTHUB_DEVICE_CLIENT_HANDLE, iotHubClientHandle, const char*, destinationFileName, IOTHUB_CLIENT_FILE_UPLOAD_GET_DATA_CALLBACK_EX, getDataCallbackEx, void*, context);
 
+     /**
+    * @brief    This API creates a context for a new blob upload to Azure Storage.
+    * @remark   It is part of a set of functions for more granular control over Azure IoT-based blob uploads.
+    *           This function is expected to be used along with:
+    *           `IoTHubDeviceClient_CreateUploadContext`
+    *           `IoTHubDeviceClient_UploadBlockToBlob`
+    *           `IoTHubDeviceClient_DestroyUploadContext`
+    *           For the standard less-granular uploads to blob please use either
+    *           `IoTHubDeviceClient_UploadToBlobAsync` or `IoTHubDeviceClient_UploadMultipleBlocksToBlobAsync`.
+    *
+    * @param    iotHubClientHandle      The handle created by a call to the create function.
+    * @param    destinationFileName     name of the file.
+    *
+    * @warning  This is a synchronous/blocking function.
+    *
+    * @return   A `IOTHUB_CLIENT_LL_UPLOADTOBLOB_CONTEXT_HANDLE` on success or NULL if the function fails.
+    */
+    MOCKABLE_FUNCTION(, IOTHUB_CLIENT_LL_UPLOADTOBLOB_CONTEXT_HANDLE, IoTHubDeviceClient_CreateUploadContext, IOTHUB_DEVICE_CLIENT_HANDLE, iotHubClientHandle, const char*, destinationFileName);
```

Comment:
Please add more details in the documentation:
1. Can the application create multiple contexts attached to the same client?
2. Since LL functions can't be called in parallel (from two different threads), what is the purpose of having multiple contexts?
3. Can I interleave block uploads? <context1(block1), context2(block1), contex1(block2)... etc>?

Reply:
I was confused for 2 as I saw the LL return type, making things a bit more confusing: are these contexts allowed in separate threads, since they are not LL functions?

Reply:
Created the types as we discussed for both iothub_device_client.h and iothub_device_client_ll.h